File: samples/fpc/trace.pas
{
  trace.pas — a complete user trace session driven through fbintf's
  Services API.

  The fbintf twin of ../cpp/trace.cpp — and proof that fbintf, unlike
  the Rust twin (rsfbclient has no Services API at all), can walk the
  same path fbtracemgr takes:

    service A: isc_action_svc_trace_start with an inline configuration
               (isc_spb_trc_cfg); the service then streams the session's
               TraceLog back line by line (isc_info_svc_line);
    worker:    a TThread attaches to the traced database and runs one
               marker query — the observed side of the stream;
    service B: isc_action_svc_trace_stop ends the session, which ends
               A's stream.

  The trace configuration targets exactly one database, so the stream
  shows only the worker's doings: ATTACH, START_TRANSACTION, the
  statement with plan and per-table counters, COMMIT, DETACH.
  See ../../trace-and-audit.md.

  Build & run (see ../README.md):
      make -C samples/fpc bin/trace && samples/fpc/bin/trace
}
program trace;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses {$IFDEF UNIX} cthreads, {$ENDIF} SysUtils, Classes, IB, FBHandsOn;

var SessionID: integer;   { written by the stream reader, read by the worker }
    OutLock: TRTLCriticalSection;

{ Output is a threadvar in FPC, one buffer per thread: serialize whole
  lines AND flush before releasing the lock, or lines interleave. }
procedure SyncWrite(const s: AnsiString);
begin
  EnterCriticalSection(OutLock);
  writeln(s);
  Flush(Output);
  LeaveCriticalSection(OutLock);
end;

function TraceConfig: AnsiString;
begin
  Result :=
    'database = ' + DbPath('trace') + #10 +
    '{'#10 +
    '  enabled = true'#10 +
    '  log_connections = true'#10 +
    '  log_transactions = true'#10 +
    '  log_statement_finish = true'#10 +
    '  print_plan = true'#10 +
    '  print_perf = true'#10 +
    '  time_threshold = 0'#10 +
    '}'#10;
end;

function AttachSvc: IServiceManager;
var SPB: ISPB;
begin
  SPB := FirebirdAPI.AllocateSPB;
  SPB.Add(isc_spb_user_name).setAsString(HandsOnUser);
  SPB.Add(isc_spb_password).setAsString(HandsOnPassword);
  Result := FirebirdAPI.GetServiceManager('localhost', TCP, SPB);
end;

{ Fetch one service output line; false when the stream is finished. }
function NextLine(Svc: IServiceManager; var line: AnsiString): boolean;
var Req: ISRB;
    Results: IServiceQueryResults;
    Item: IServiceQueryResultItem;
begin
  Req := Svc.AllocateSRB;
  Req.Add(isc_info_svc_line);
  Results := Svc.Query(nil, Req);
  Item := Results.find(isc_info_svc_line);
  Result := (Item <> nil);
  if Result then
  begin
    line := Item.getAsString;
    Result := line <> '';
  end;
end;

type
  { The observed side, then the stop from a second service. }
  TWorker = class(TThread)
  protected
    procedure Execute; override;
  end;

procedure TWorker.Execute;
var Att: IAttachment;
    Tr: ITransaction;
    SvcB: IServiceManager;
    Req: ISRB;
    line: AnsiString;
begin
  Sleep(800);
  Att := FirebirdAPI.OpenDatabase(DbConn('trace'), DefaultDPB);
  Tr := Att.StartTransaction([isc_tpb_read, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  SyncWrite('[worker] marker query says: ' + Trim(Att.OpenCursorAtStart(Tr,
    'SELECT COUNT(*) FROM RDB$RELATIONS /* traced! */')[0].AsString));
  Tr.Commit;
  Att := nil;                             { DETACH, while the session runs }

  Sleep(1200);
  while SessionID = 0 do                  { wait for the id from the stream }
    Sleep(50);
  SvcB := AttachSvc;
  Req := SvcB.AllocateSRB;
  Req.Add(isc_action_svc_trace_stop);
  Req.Add(isc_spb_trc_id).SetAsInteger(SessionID);
  SvcB.Start(Req);
  while NextLine(SvcB, line) do
    SyncWrite('[stop ] ' + TrimRight(line));
  SvcB.Detach;
end;

var Att: IAttachment;
    SvcA: IServiceManager;
    Req: ISRB;
    Worker: TWorker;
    line: AnsiString;
    p: integer;

begin
  SessionID := 0;
  InitCriticalSection(OutLock);

  { The database to be observed must exist before the session starts. }
  Att := AttachOrCreate(DbConn('trace'));
  Att := nil;

  { -- service A: start the trace session, config passed inline ------- }
  SvcA := AttachSvc;
  Req := SvcA.AllocateSRB;
  Req.Add(isc_action_svc_trace_start);
  Req.Add(isc_spb_trc_name).SetAsString('hands-on');
  Req.Add(isc_spb_trc_cfg).SetAsString(TraceConfig);
  SvcA.Start(Req);

  Worker := TWorker.Create(false);

  { -- service A's stream: the trace output itself -------------------- }
  while NextLine(SvcA, line) do
  begin
    SyncWrite('[trace] ' + TrimRight(line));
    if (SessionID = 0) and (Pos('Trace session ID ', line) = 1) then
      if TryStrToInt(Trim(Copy(line, Length('Trace session ID ') + 1,
          Pos(' started', line) - Length('Trace session ID ') - 1)), p) then
        SessionID := p;
  end;

  Worker.WaitFor;
  Worker.Free;
  SvcA.Detach;
  DoneCriticalSection(OutLock);
  writeln('done.');
end.

File: samples/fb-cpp/trace.cpp
/*
 *  trace.cpp (fb-cpp) — a live trace session, typed where fb-cpp reaches.
 *
 *  The fb-cpp twin of ../cpp/trace.cpp: service A starts a user trace
 *  session with an inline configuration and streams its TraceLog back; a
 *  worker attaches (its own Client + Attachment, fb-cpp's per-thread idiom)
 *  and runs one marker query — the observed side; service B stops the
 *  session by id.  fb-cpp wraps service ACTIONS (backup, restore, repair),
 *  not the trace family, so this twin shows the layered escape hatch at
 *  full stretch: ServiceManager owns the attach/detach lifecycle (SPB_ATTACH
 *  built from typed options, RAII detach), while trace_start/query/trace_stop
 *  drop to the underlying IService via getHandle() — the same raw tags as
 *  the OO-API version, inside an otherwise typed program.
 *  See ../../trace-and-audit.md.
 *
 *  Build & run (see ../README.md):
 *      ./build/fbcpp_trace
 */

#include "fbcpp_sample.h"
#include <ibase.h>
#include <atomic>
#include <chrono>
#include <cstdio>
#include <string>
#include <thread>

using namespace fbcpp;
using namespace fbcpp_sample;

static const char* DB_PATH = "/tmp/fbhandson/trace_fbcpp.fdb";

static const char* TRACE_CFG =
	"database = /tmp/fbhandson/trace_fbcpp.fdb\n"
	"{\n"
	"  enabled = true\n"
	"  log_connections = true\n"
	"  log_transactions = true\n"
	"  log_statement_finish = true\n"
	"  print_plan = true\n"
	"  print_perf = true\n"
	"  time_threshold = 0\n"
	"}\n";

static ServiceManagerOptions svcOptions()
{
	return ServiceManagerOptions()
		.setServer("localhost")
		.setUserName(envOr("ISC_USER", "SYSDBA"))
		.setPassword(envOr("ISC_PASSWORD", "masterkey"));
}

// Fetch one service output line; returns false when the stream is finished.
static bool nextLine(fb::CheckStatusWrapper& st, fb::IService* svc, std::string& line)
{
	const unsigned char items[] = { isc_info_svc_line };
	unsigned char buf[4096];
	svc->query(&st, 0, nullptr, sizeof(items), items, sizeof(buf), buf);
	if ((st.getState() & fb::IStatus::STATE_ERRORS) || buf[0] != isc_info_svc_line)
		return false;
	const unsigned len = buf[1] | (unsigned(buf[2]) << 8);
	line.assign(reinterpret_cast<char*>(buf + 3), len);
	return len != 0;
}

int main()
{
	try
	{
		Client client{"fbclient"};

		// The database to be observed must exist before the session starts.
		attachOrCreate(client, std::string("inet://localhost/") + DB_PATH);

		fb::IMaster* master = client.getMaster();
		fb::IUtil* utl = master->getUtilInterface();

		// -- service A: typed attach, then trace_start through the handle ---
		ServiceManager svcA{client, svcOptions()};
		fb::CheckStatusWrapper st{master->getStatus()};
		fb::IXpbBuilder* start = utl->getXpbBuilder(&st, fb::IXpbBuilder::SPB_START, nullptr, 0);
		start->insertTag(&st, isc_action_svc_trace_start);
		start->insertString(&st, isc_spb_trc_name, "hands-on (fb-cpp)");
		start->insertString(&st, isc_spb_trc_cfg, TRACE_CFG);
		svcA.getHandle()->start(&st, start->getBufferLength(&st), start->getBuffer(&st));
		start->dispose();
		if (st.getState() & fb::IStatus::STATE_ERRORS)
			throw FbCppException("trace_start failed");

		std::atomic<int> sessionId{0};

		// -- worker: the observed side, then the stop from a second service -
		std::thread worker([&sessionId]()
		{
			std::this_thread::sleep_for(std::chrono::milliseconds(800));
			try
			{
				Client c{"fbclient"};
				Attachment att{c, std::string("inet://localhost/") + DB_PATH,
					defaultOptions()};
				Transaction tra{att};
				Statement marker{att, tra,
					"SELECT COUNT(*) FROM RDB$RELATIONS /* traced! */"};
				marker.execute(tra);
				printf("[worker] marker query says: %s\n",
					marker.getString(0)->c_str());
				tra.commit();
			}
			catch (const std::exception& e) { report(e); }

			std::this_thread::sleep_for(std::chrono::milliseconds(1200));
			while (!sessionId.load())                   // wait for the id from the stream
				std::this_thread::sleep_for(std::chrono::milliseconds(50));
			try
			{
				Client c{"fbclient"};
				ServiceManager svcB{c, svcOptions()};
				fb::CheckStatusWrapper st2{c.getMaster()->getStatus()};
				fb::IUtil* utl2 = c.getMaster()->getUtilInterface();
				fb::IXpbBuilder* stop = utl2->getXpbBuilder(&st2, fb::IXpbBuilder::SPB_START, nullptr, 0);
				stop->insertTag(&st2, isc_action_svc_trace_stop);
				stop->insertInt(&st2, isc_spb_trc_id, sessionId.load());
				svcB.getHandle()->start(&st2, stop->getBufferLength(&st2), stop->getBuffer(&st2));
				stop->dispose();
				std::string line;
				while (nextLine(st2, svcB.getHandle().get(), line))
					printf("[stop ] %s\n", line.c_str());
				st2.dispose();
			}
			catch (const std::exception& e) { report(e); }
		});

		// -- service A's stream: the trace output itself --------------------
		std::string line;
		while (nextLine(st, svcA.getHandle().get(), line))
		{
			printf("[trace] %s\n", line.c_str());
			int id;
			if (sscanf(line.c_str(), "Trace session ID %d started", &id) == 1)
				sessionId.store(id);
		}

		worker.join();
		st.dispose();
		printf("done.\n");
		return 0;
	}
	catch (const std::exception& e)
	{
		return report(e);
	}
}
